joern-cli/frontends/javasrc2cpg/src/main/scala/io/joern/javasrc2cpg/passes/AstCreator.scala

```diff
@@ -2323,183 +2926,353 @@ class AstCreator(filename: String, javaParserAst: CompilationUnit, global: Globa
-    val refEdgePairs    = buildRefEdgePairs(identifiersMatchingParams, namesToMethodParams)
+    val parameterNodes = lambdaParameters
+      .lazyZip(paramTypesList)
+      .lazyZip(LazyList.from(1))
+      .map { case (param, typ, idx) =>
+        val name = param.getNameAsString
+        NewMethodParameterIn()
+          .name(name)
+          .typeFullName(typ)
+          .order(idx)
+          .code(s"$typ $name")
+          .evaluationStrategy(EvaluationStrategies.BY_SHARING)
+          .lineNumber(line(expr))
+      }
 
-    val methodAst =
-      lambdaMethodAst(expr, fullName, parameterAsts, bodyAst, closureBindings, refEdgePairs)
+    parameterNodes.foreach { paramNode =>
+      scopeStack.addToScope(paramNode.name, paramNode)
+    }
 
-    val methodRef = lambdaMethodRef(expr, fullName, order)
+    parameterNodes.map(Ast(_))
+  }
 
-    val closuresWithMeta = buildClosuresWithMeta(closureBindings, methodRef)
-    lambdaContextQueue.append(Context(lambdaAsts = Seq(methodAst), closureBindingInfo = closuresWithMeta))
+  private def getLambdaReturnType(
+    maybeResolvedLambdaType: Option[ResolvedType],
+    maybeBoundMethod: Option[ResolvedMethodDeclaration],
+    expectedTypeParamTypes: Map[ResolvedTypeParameterDeclaration, ResolvedType]
+  ): String = {
+    maybeBoundMethod
+      .map { boundMethod =>
+        boundMethod.getReturnType match {
+          case returnType: ResolvedTypeVariable =>
+            expectedTypeParamTypes.getOrElse(returnType.asTypeParameter, returnType)
 
-    Ast(methodRef)
+          case returnType => returnType
+        }
+      }
+      .orElse {
+        maybeResolvedLambdaType
+      }
+      .map {
+        typeInfoCalc.fullName
+      }
+      .getOrElse(TypeConstants.UnresolvedType)
   }
 
-  private def buildClosuresWithMeta(
-    closureBindings: Seq[ClosureBindingInfo],
-    methodRef: NewMethodRef
-  ): Seq[ClosureBindingMeta] = {
-    // This is just to get javasrc to compile. Lambdas are completeley broken, so this will probably be rewritten
-    // completely
-    val nameToNode: Map[String, NewNode] = Map.empty
-
-    closureBindings.map { closureBindingInfo =>
-      val name  = closureBindingInfo.identifier.name
-      val edges = List((methodRef, closureBindingInfo.closure, EdgeTypes.CAPTURE))
-
-      val refEdges =
-        if (nameToNode.contains(name)) {
-          val node = nameToNode(name)
-          List((closureBindingInfo.closure, node, EdgeTypes.REF))
+  def closureBinding(closureBindingId: String, originalName: String): NewClosureBinding = {
+    NewClosureBinding()
+      .closureBindingId(closureBindingId)
+      .closureOriginalName(originalName)
+      .evaluationStrategy(EvaluationStrategies.BY_SHARING)
+  }
+
+  private def closureBindingsForCapturedNodes(captured: List[NamedVariableNodeType]): List[ClosureBindingEntry] = {
+    captured
+      .map { capturedNode =>
+        val closureBindingId   = randomUUID().toString
```

Comment:
add a TODO that this should not be random. we want a CPG generated from the same source to be identical on a new run